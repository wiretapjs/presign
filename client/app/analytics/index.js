import dispatchHeapIdentification from 'lib/services/analytics/heap/dispatchIdentification'
import dispatchHeapAddUserProperties from 'lib/services/analytics/heap/dispatchAddUserProperties'
import analyticsManager from 'lib/services/analytics/analyticsManager'
import AnalyticsEvent from 'lib/services/analytics/AnalyticsEvent'

export const identifyUser = function (identificationObject) {
  if (!identificationObject) return
  const {emailAddress, policyNumber} = identificationObject
  if (emailAddress) {
    analyticsManager.analyticsEventsArr.push(
      new AnalyticsEvent(emailAddress, dispatchHeapIdentification))
  }

  if (policyNumber) {
    analyticsManager.analyticsEventsArr.push(
      new AnalyticsEvent({policyNumber}, dispatchHeapAddUserProperties))
  }
}
