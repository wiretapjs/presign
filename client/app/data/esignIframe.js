export const events = Object.freeze({
  REGISTER: 'ESL:MESSAGE:REGISTER',
  ACTIVATE_EVENTS: 'ESL:MESSAGE:ACTIVATE_EVENTS',
  DOCUMENT_ACCEPT_STARTED: 'ESL:MESSAGE:STARTED:DOCUMENT_ACCEPT',
  DOCUMENT_ACCEPT_SUCCESS: 'ESL:MESSAGE:SUCCESS:DOCUMENT_ACCEPT',
  DOCUMENT_ACCEPT_ERROR: 'ESL:MESSAGE:ERROR:DOCUMENT_ACCEPT',
  PACKAGE_OPT_OUT_STARTED: 'ESL:MESSAGE:STARTED:PACKAGE_OPT_OUT',
  PACKAGE_OPT_OUT_SUCCESS: 'ESL:MESSAGE:SUCCESS:PACKAGE_OPT_OUT',
  PACKAGE_OPT_OUT_ERROR: 'ESL:MESSAGE:ERROR:PACKAGE_OPT_OUT',
  PACKAGE_DECLINE_STARTED: 'ESL:MESSAGE:STARTED:PACKAGE_DECLINE',
  PACKAGE_DECLINE_SUCCESS: 'ESL:MESSAGE:SUCCESS:PACKAGE_DECLINE',
  PACKAGE_DECLINE_ERROR: 'ESL:MESSAGE:ERROR:PACKAGE_DECLINE',
  DOCUMENT_CONFIRM_STARTED: 'ESL:MESSAGE:STARTED:DOCUMENT_CONFIRM',
  DOCUMENT_CONFIRM_SUCCESS: 'ESL:MESSAGE:SUCCESS:DOCUMENT_CONFIRM',
  DOCUMENT_CONFIRM_ERROR: 'ESL:MESSAGE:ERROR:DOCUMENT_CONFIRM',
  SIGNER_COMPLETE_STARTED: 'ESL:MESSAGE:STARTED:SIGNER_COMPLETE',
  SIGNER_COMPLETE_SUCCESS: 'ESL:MESSAGE:SUCCESS:SIGNER_COMPLETE',
  DOCUMENT_NAVIGATION_STARTED: 'ESL:MESSAGE:STARTED:DOCUMENT_NAVIGATION',
  DOCUMENT_NAVIGATION_SUCCESS: 'ESL:MESSAGE:SUCCESS:DOCUMENT_NAVIGATION',
})

export const errorEvents = [
  events.DOCUMENT_ACCEPT_ERROR,
  events.DOCUMENT_CONFIRM_ERROR,
  events.PACKAGE_DECLINE_ERROR,
  events.PACKAGE_OPT_OUT_ERROR,
]

export const signerCompleteEvents = [
  events.SIGNER_COMPLETE_STARTED,
  events.SIGNER_COMPLETE_SUCCESS,
]
