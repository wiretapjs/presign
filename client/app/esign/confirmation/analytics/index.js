import AnalyticsEvent from 'lib/services/analytics/AnalyticsEvent'
import dispatchOmniture from 'lib/services/analytics/omniture/dispatch'
import analyticsManager from 'lib/services/analytics/analyticsManager'
import OmniturePageView from './page-view/OmniturePayload'

export const recordEsignConfirmationView = function (policyNumber, jurisdiction) {
  analyticsManager.analyticsEventsArr.push(
    new AnalyticsEvent(new OmniturePageView(policyNumber, jurisdiction), dispatchOmniture),
  )
}
