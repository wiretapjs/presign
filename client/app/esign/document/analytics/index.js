import analyticsManager from 'lib/services/analytics/analyticsManager'
import AnalyticsEvent from 'lib/services/analytics/AnalyticsEvent'
import dispatchOmniture from 'lib/services/analytics/omniture/dispatch'
import heapConfig from 'lib/services/analytics/heap/config'
import dispatchHeapTracking from 'lib/services/analytics/heap/dispatchTracking'
import OmniturePayload from './page-view/OmniturePayload'
import OmnitureErrorPayload from './page-view/OmnitureErrorPayload'

const isESignDocumentDataMissing = function (eSignSession, jurisdiction) {
  return !eSignSession || !jurisdiction
}

export const recordESignDocumentError = function () {
  analyticsManager.analyticsEventsArr
    .push(new AnalyticsEvent(new OmnitureErrorPayload(), dispatchOmniture))
}

export const recordESignDocument = function (eSignSession, jurisdiction) {
  if (isESignDocumentDataMissing(eSignSession, jurisdiction)) {
    recordESignDocumentError()
  } else {
    analyticsManager.analyticsEventsArr
      .push(new AnalyticsEvent(new OmniturePayload(eSignSession, jurisdiction), dispatchOmniture))
  }
}

export const recordEsignCantSeeDocumentClick = function () {
  analyticsManager.analyticsEventsArr
    .push(new AnalyticsEvent(heapConfig.eSignDocumentScreen.cantSeeDocsLink, dispatchHeapTracking))
}

export const recordNeedHelpClick = function () {
  analyticsManager.analyticsEventsArr
    .push(new AnalyticsEvent(heapConfig.eSignDocumentScreen.needHelp, dispatchHeapTracking))
}
