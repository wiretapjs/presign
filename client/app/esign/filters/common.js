import filters from 'app/esign/filters/sessions/filterCollection'

export const getDocumentCountByDocType = function (sessions, docType) {
  return sessions.reduce((total, policy) => {
    const sum = policy.eSignSession.reduce((accumulator, session) => {
      const count = +session[docType.countField] || 0
      return accumulator + count
    }, 0)

    return total + sum
  }, 0)
}

export const canEsign = function (sessions) {
  let result = true

  filters.forEach(filter => {
    const filterResult = filter(sessions)
    if (!filterResult) {
      result = false
    }
  },
  )

  return result
}
