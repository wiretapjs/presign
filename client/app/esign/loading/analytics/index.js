import AnalyticsEvent from 'lib/services/analytics/AnalyticsEvent'
import dispatchOmniture from 'lib/services/analytics/omniture/dispatch'
import analyticsManager from 'lib/services/analytics/analyticsManager'
import OmniturePayload from './page-view/OmniturePayload'
import OmnitureErrorPayload from './page-view/OmnitureErrorPayload'

export const recordESignLoadingError = function () {
  analyticsManager.analyticsEventsArr
    .push(new AnalyticsEvent(new OmnitureErrorPayload(), dispatchOmniture))
}

export const recordESignLoading = function (eSignSession) {
  if (!eSignSession) {
    recordESignLoadingError()
  } else {
    analyticsManager.analyticsEventsArr
      .push(new AnalyticsEvent(new OmniturePayload(eSignSession), dispatchOmniture))
  }
}
