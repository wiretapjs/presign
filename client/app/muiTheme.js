import getMuiTheme from 'material-ui/styles/getMuiTheme'

const theme = {
  raisedButton: {
    color: '#037BB7',
    textColor: '#FFFFFF',
    disabledColor: '#E5E5E5',
    disabledTextColor: '#A3A3A3',
  },
}

export default getMuiTheme(theme)
