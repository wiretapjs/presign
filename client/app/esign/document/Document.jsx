import React, { Component } from 'react'
import PropTypes from 'prop-types'
import sessionsService from 'lib/services/sessions'
import { EsignIframe, Footer, Header, MainTitle } from 'lib/components'
import ceremoniesService from 'lib/services/ceremonies'
import policyService from 'lib/services/policies'
import autoBind from 'react-autobind'
import { errorEvents, signerCompleteEvents } from 'app/data/esignIframe'
import View from 'app/View'
import routing from 'lib/util/routing'
import redirect from 'lib/util/redirect'
import isURL from 'validator/lib/isURL'
import { canEsign } from 'app/esign/filters/common'
import RegistrationLoading from 'app/registration/RegistrationLoading'
import HelpLink from './HelpLink'
import { recordESignDocument, recordESignDocumentError } from './analytics'
import CantSeeDocsLinkComponent from './CantSeeDocsLinkComponent'
import {
  StyledHeaderWrapper,
  StyledWrapper,
  StyledIframeWrapper,
  StyledTextWrapper,
  StyledErrorMessage,
  StyledHelpWrapper } from './Document.style'

class Document extends Component {
  constructor () {
    super()
    this.state = {
      esignatureUrl: '',
      policyNumber: '',
      showLoading: true,
      showErrorMessage: false,
    }
    autoBind(this)
  }

  async componentDidMount () {
    try {
      const sessionList = await sessionsService.getUpdatedSessions()

      if (!canEsign(sessionList)) {
        this.handleLoadingError()
      } else {
        const sessionObjectForFirstPolicy = sessionList[0]
        const policyNumber = sessionObjectForFirstPolicy.policyNumber

        const firstEsignSession = sessionObjectForFirstPolicy.eSignSession[0]
        const firstSignerOfFirstSession = firstEsignSession.signerSession[0]
        const esignatureUrl = firstSignerOfFirstSession.esignatureUrl
        if (!isURL(esignatureUrl)) throw new Error('Invalid URL')
        const policySummary = await policyService.getCachedSummary(policyNumber)

        recordESignDocument(sessionObjectForFirstPolicy, policySummary.policyJurisdiction)

        this.setState({esignatureUrl, policyNumber, showLoading: false})
      }
    } catch (err) {
      this.handleLoadingError()
    }
  }

  handleLoadingError () {
    recordESignDocumentError()
    this.redirectToEsign()
  }

  redirectToEsign () {
    redirect.eserviceEsignLogin()
  }

  esignEventsHandler (data, domain) {
    if (signerCompleteEvents.includes(data)) {
      this.validateCeremoniesAndRedirect()
    } else if (errorEvents.includes(data)) {
      this.setState({showErrorMessage: true})
      this.redirectToEsign()
    }
  }

  validateCeremoniesAndRedirect () {
    this.setState({showLoading: true})
    this.areCeremoniesValid().then((areValid) => {
      if (areValid) {
        this.props.history.push(routing.generateLinkTo('/esign/confirmation'))
      } else {
        this.redirectToEsign()
      }
    })
  }

  areCeremoniesValid () {
    return ceremoniesService.getCeremonies({
      policyNumbers: [this.state.policyNumber],
      sessionType: 'esign',
    }).then((ceremoniesResponse) => {
      const { ceremonies } = ceremoniesResponse
      if (!ceremonies || !ceremonies.length) return false

      return ceremonies[0].ceremonyStatus !== 'failed' && ceremonies[0].redirectIndicator !== 'Y'
    }).catch(() => false)
  }

  renderIframe () {
    if (this.state.showLoading) {
      return <RegistrationLoading />
    }

    if (this.state.showErrorMessage) {
      return (<StyledErrorMessage>Oops! It looks like something went wrong.</StyledErrorMessage>)
    }

    return <EsignIframe url={this.state.esignatureUrl} esignEvents={this.esignEventsHandler} />
  }

  render () {
    return (
      <View>
        <StyledWrapper>
          <StyledHeaderWrapper>
            <Header />
          </StyledHeaderWrapper>
          <StyledTextWrapper>
            <MainTitle> Let's sign your policy documents </MainTitle>
          </StyledTextWrapper>
          <StyledHelpWrapper>
            <CantSeeDocsLinkComponent />
            <HelpLink />
          </StyledHelpWrapper>
          <StyledIframeWrapper>
            {this.renderIframe()}
          </StyledIframeWrapper>
          <Footer />
        </StyledWrapper>
      </View>
    )
  }
}

Document.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
}

export default Document
