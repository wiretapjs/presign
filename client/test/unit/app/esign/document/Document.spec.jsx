import React from 'react'
import { mount, shallow } from 'enzyme'
import _ from 'lodash'
import { context } from 'test/helpers/jest'
import muiTheme from 'app/muiTheme'
import ESignDocumentView from 'app/esign/document/Document'
import {StyledErrorMessage} from 'app/esign/document/Document.style'
import sessionsService from 'lib/services/sessions'
import redirect from 'lib/util/redirect'
import * as common from 'app/esign/filters/common'
import ceremoniesService from 'lib/services/ceremonies'
import policyService from 'lib/services/policies'
import * as esignIframeData from 'app/data/esignIframe'
import { flushPromises } from 'test/helpers/promises'
import { recordESignDocument, recordESignDocumentError } from 'app/esign/document/analytics'

const history = {
  push: jest.fn(),
}

const mountWithContext = node => mount(node, {
  context: { muiTheme },
  childContextTypes: { muiTheme: React.PropTypes.object },
})

redirect.eserviceEsignLogin = jest.fn()
let esignRenderWrapper
let esignDocumentInstance

jest.mock('app/esign/document/analytics')

const sessionsResponse = [{
  eSignSession: [
    {
      signerSession: [
        { namedInsuredNumber: 1,
          esignatureUrl: 'http://example.com',
        },
      ],
      numSignOnlineDocs: '3',
      numPrintSignMailDocs: '0',
    },
  ],
  signerCount: 1,
  policyNumber: 5,
},
]

describe('Esign Document View', () => {
  beforeEach(() => {
    ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
      ceremonies: [
        {
          redirectIndicator: 'N',
          ceremonyStatus: 'complete',
        },
      ],
    }))
    esignRenderWrapper = shallow(<ESignDocumentView history={history} />)
    esignDocumentInstance = esignRenderWrapper.instance()
    common.canEsign = jest.fn(() => true)
    sessionsService.getUpdatedSessions = jest.fn(() => Promise.resolve(
      sessionsResponse,
    ))
    policyService.getCachedSummary = jest.fn(() => Promise.resolve(
      {
        policyNumber: 12345,
        policyJurisdiction: 'CA',
      },
    ))
  })
  afterEach(() => {
    jest.resetAllMocks()
  })

  it('it should render Esign Document View', () => {
    expect(esignRenderWrapper).toMatchSnapshot()
  })
  it('it should render Error message if isErrorMessage is true', () => {
    esignRenderWrapper = mountWithContext(
      <ESignDocumentView history={history} />,
    )
    esignRenderWrapper.setState({showErrorMessage: true, showLoading: false})
    expect(esignRenderWrapper.find(StyledErrorMessage).length).toBe(1)
  })

  describe('on mount', () => {
    it('should set show loading to false once get the list of esign sessions and the policy summary', async () => {
      await esignDocumentInstance.componentDidMount()
      expect(sessionsService.getUpdatedSessions).toHaveBeenCalledTimes(1)
      expect(policyService.getCachedSummary).toHaveBeenCalledTimes(1)
      expect(esignRenderWrapper.state().showLoading).toBe(false)
    })

    context('with eSign sessions that meet E2 filter criteria', () => {
      beforeEach(async () => {
        await esignDocumentInstance.componentDidMount()
      })

      it('should record the page load analytics event', () => {
        expect(recordESignDocument).toHaveBeenCalledTimes(1)
      })

      it('it should not redirect', async () => {
        await esignDocumentInstance.componentDidMount()
        expect(redirect.eserviceEsignLogin).not.toHaveBeenCalled()
      })
    })

    context('with an invalid esignature url', () => {
      beforeEach(async () => {
        const sessionsResponseWithInvalidUrl = _.cloneDeep(sessionsResponse)
        sessionsResponseWithInvalidUrl[0].eSignSession[0].signerSession[0].esignatureUrl = 'false'
        sessionsService.getUpdatedSessions = jest.fn(
          () => Promise.resolve(sessionsResponseWithInvalidUrl))
        await esignDocumentInstance.componentDidMount()
        await flushPromises()
      })
      it('should record the error analytics event', () => {
        expect(recordESignDocumentError).toHaveBeenCalledTimes(1)
      })

      it('it should redirect', () => {
        expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
      })
    })

    context('with an error response from the eSign session service', () => {
      beforeEach(async () => {
        sessionsService.getUpdatedSessions = jest.fn(() => Promise.reject(new Error()))
        await esignDocumentInstance.componentDidMount()
        await flushPromises()
      })

      it('should record the error analytics event', () => {
        expect(recordESignDocumentError).toHaveBeenCalledTimes(1)
      })

      it('it should redirect', () => {
        expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
      })
    })

    context('with eSign sessions that do not meet E2 filter criteria', () => {
      beforeEach(async () => {
        common.canEsign = jest.fn(() => false)
        await esignDocumentInstance.componentDidMount()
      })

      it('it should redirect', () => {
        expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
      })

      it('should record the error analytics event', () => {
        expect(recordESignDocumentError).toHaveBeenCalledTimes(1)
      })
    })

    context('with an error response from the policy service', () => {
      beforeEach(async () => {
        policyService.getCachedSummary = jest.fn(() => Promise.reject(new Error()))
        await esignDocumentInstance.componentDidMount()
        await flushPromises()
      })

      it('should redirect', () => {
        expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
      })

      it('should record the error analytics event', () => {
        expect(recordESignDocumentError).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('esignEventsHandler', () => {
    const domain = 'libertymutual.com'
    beforeEach(async () => {
      await esignDocumentInstance.componentDidMount()
      await flushPromises()
    })

    context('with a SIGNER_COMPLETE_STARTED event', () => {
      const event = esignIframeData.events.SIGNER_COMPLETE_STARTED

      it('does not show an error message', () => {
        esignDocumentInstance.esignEventsHandler(event, domain)
        expect(esignDocumentInstance.state.showErrorMessage).toBe(false)
      })

      it('sets showLoading to true', () => {
        esignDocumentInstance.esignEventsHandler(event, domain)
        expect(esignDocumentInstance.state.showLoading).toBe(true)
      })

      context('with a valid ceremonies result', () => {
        it('redirects to the esign confirmation page', async () => {
          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(history.push).toHaveBeenCalledWith({
            pathname: '/esign/confirmation',
            search: window.location.search,
          })
        })
      })

      context('with an empty ceremonies result', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
            ceremonies: [],
          }))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with an empty result', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({}))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a ceremony with a redirectIndicator', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
            ceremonies: [
              {
                redirectIndicator: 'Y',
                ceremonyStatus: 'complete',
              },
            ],
          }))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a ceremony with a ceremonyStatus', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
            ceremonies: [
              {
                redirectIndicator: 'Y',
                ceremonyStatus: 'failed',
              },
            ],
          }))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a failing ceremonies service call', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.reject(new Error()))
          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })
    })

    context('with a SIGNER_COMPLETE_SUCCESS', () => {
      const event = esignIframeData.events.SIGNER_COMPLETE_SUCCESS

      it('does not show an error message', () => {
        esignDocumentInstance.esignEventsHandler(event, domain)
        expect(esignDocumentInstance.state.showErrorMessage).toBe(false)
      })

      it('sets showLoading to true', () => {
        esignDocumentInstance.esignEventsHandler(event, domain)
        expect(esignDocumentInstance.state.showLoading).toBe(true)
      })

      context('with a valid ceremonies result', () => {
        it('redirects to the esign confirmation page', async () => {
          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(history.push).toHaveBeenCalledWith({
            pathname: '/esign/confirmation',
            search: window.location.search,
          })
        })
      })

      context('with an empty ceremonies result', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
            ceremonies: [],
          }))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with an empty result', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({}))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a ceremony with a ceremonyStatus', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.resolve({
            ceremonies: [
              {
                redirectIndicator: 'Y',
                ceremonyStatus: 'failed',
              },
            ],
          }))

          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a failing ceremonies service call', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.reject(new Error()))
          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })

      context('with a failing ceremonies service call', () => {
        it('redirects to the yellow box', async () => {
          ceremoniesService.getCeremonies = jest.fn(() => Promise.reject(new Error()))
          esignDocumentInstance.esignEventsHandler(event, domain)
          await flushPromises()
          expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
        })
      })
    })

    context('with another non-error event', () => {
      beforeEach(() => {
        esignDocumentInstance.esignEventsHandler('foo', domain)
      })

      it('does not show an error message', () => {
        expect(esignDocumentInstance.state.showErrorMessage).toBe(false)
      })

      it('does not set showLoading to true', () => {
        expect(esignDocumentInstance.state.showLoading).toBe(false)
      })
    })

    context('with an error event', () => {
      beforeEach(() => {
        const data = esignIframeData.events.PACKAGE_DECLINE_ERROR
        esignDocumentInstance.esignEventsHandler(data, domain)
      })

      it('does not set showLoading to true', () => {
        expect(esignDocumentInstance.state.showLoading).toBe(false)
      })

      it('sets showError to true', () => {
        expect(esignDocumentInstance.state.showErrorMessage).toBe(true)
      })
      it('redirect to esign', () => {
        expect(redirect.eserviceEsignLogin).toHaveBeenCalledTimes(1)
      })
    })

    context('with an undefined event', () => {
      beforeEach(() => {
        esignDocumentInstance.esignEventsHandler(undefined, domain)
      })

      it('does not show an error message', () => {
        expect(esignDocumentInstance.state.showErrorMessage).toBe(false)
      })

      it('does not set showLoading to true', () => {
        expect(esignDocumentInstance.state.showLoading).toBe(false)
      })
    })
  })
})
