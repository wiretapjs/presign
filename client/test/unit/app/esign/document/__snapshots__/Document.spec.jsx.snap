// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Esign Document View it should render Esign Document View 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <View>
    <styled.div>
      <styled.div>
        <Header
          leftPadding="90px"
        />
      </styled.div>
      <styled.div>
        <styled.h1>
           Let's sign your policy documents 
        </styled.h1>
      </styled.div>
      <styled.div>
        <CantSeeDocsLinkComponent />
        <HelpLink />
      </styled.div>
      <styled.div>
        <RegistrationLoading />
      </styled.div>
      <Footer />
    </styled.div>
  </View>,
  "nodes": Array [
    <View>
      <styled.div>
        <styled.div>
          <Header
            leftPadding="90px"
          />
        </styled.div>
        <styled.div>
          <styled.h1>
             Let's sign your policy documents 
          </styled.h1>
        </styled.div>
        <styled.div>
          <CantSeeDocsLinkComponent />
          <HelpLink />
        </styled.div>
        <styled.div>
          <RegistrationLoading />
        </styled.div>
        <Footer />
      </styled.div>
    </View>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Document
        history={
          Object {
            "push": [Function],
          }
        }
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Document {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "history": Object {
            "push": [Function],
          },
        },
        "refs": Object {},
        "state": Object {
          "esignatureUrl": "",
          "policyNumber": "",
          "showErrorMessage": false,
          "showLoading": true,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <View>
          <styled.div>
            <styled.div>
              <Header
                leftPadding="90px"
              />
            </styled.div>
            <styled.div>
              <styled.h1>
                 Let's sign your policy documents 
              </styled.h1>
            </styled.div>
            <styled.div>
              <CantSeeDocsLinkComponent />
              <HelpLink />
            </styled.div>
            <styled.div>
              <RegistrationLoading />
            </styled.div>
            <Footer />
          </styled.div>
        </View>,
        "_debugID": 2,
        "_renderedOutput": <View>
          <styled.div>
            <styled.div>
              <Header
                leftPadding="90px"
              />
            </styled.div>
            <styled.div>
              <styled.h1>
                 Let's sign your policy documents 
              </styled.h1>
            </styled.div>
            <styled.div>
              <CantSeeDocsLinkComponent />
              <HelpLink />
            </styled.div>
            <styled.div>
              <RegistrationLoading />
            </styled.div>
            <Footer />
          </styled.div>
        </View>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Document
    history={
      Object {
        "push": [Function],
      }
    }
  />,
}
`;
